src/gui/sdlmain.cpp

```diff
@@ -264,6 +264,15 @@ struct SDL_Block {
 		} window;
 		Bit8u bpp = 0;
 		bool fullscreen = false;
+		// This flag indicates, that we are in the process of switching
+		// between fullscreen or window (as oppososed to changing
+		// rendering size due to rotating screen, emulation state, or
+		// user resizing the window).
+		bool switching_fullscreen = false;
+		// Lazy window size init triggers updating window size and
+		// position when leaving fullscreen for the first time.
+		// See FinalizeWindowState function for details.
+		bool lazy_init_window_size = false;
```

Comment:
Being able to defer (and eat) video-mode changes will be very useful!

I've been wrestling with instant launch (which reorders most of the config-file parsing and command-line handling ahead of setting the first mode to show the splash), and in some cases we make up to 4 video mode changes before the program launches and the final video mode is locked in; and the majority of that wait-time is sunk into the SDL library making those changes. 

Lazy-init may let us skip these initial mode changes (even when executing internal commands in a batch file), until we hit the first "external executable binary".

Reply:
After reading, tweaking and modifying this code for some time now, I think I have a pretty good understanding of how `sdlmain` code could be redesigned to make it less coupled (and hopefully much simpler in the process). But I don't want to do any radical changes before 0.75 release (risk of regressions is extremely high here).

I think the big problem in sdlmain is the "mis-feature" of changing the output in runtime. Old code used it for drawing splash, but it is still necessary for Mapper GUI and "error dialogs" on Windows (still never seen one). It adds a lot of complexity. Removal of this option would speed up mode changes considerably (and probably avoid some bugs).

Other features, that are not really that helpful, but very hard to handle correctly and add a lot of maintenance and testing burden:

- `output=surface` - it's ridiculous, and gotos in code are very annoying. I managed to lock up my PC completely several times by going into infinite loop of dosbox trying to create textured window, falling back to surface, then trying out texture again, and again… `texture_renderer
- `windowresolution=original` - it would be much simpler if windowsize was being controlled only by explicit user actions and by OS (and not by the program itself). I would like to remove this option, but only after window resizing will work for all available outputs.
- `fullresolution=<custom>` - when user selects non-native resolution, it works terribly (based on testing I mentioned above). I am not surprised, that all modern OSes and user environments make it hardor (or even impossible) to support.
- GFX_* API is lacking - that's why I needed to add those two booleans to track internal state (instead of properly routing information through API calls).

There's also `priority` option in sdl section, which… I don't think is really needed any more? I've never had any urge to change the default, and users could get better results by setting niceness and process affinity from outside.

Reply:
> I think the big problem in sdlmain is the "mis-feature" of changing the output in runtime. Old code used it for drawing splash, but it is still necessary for Mapper GUI and "error dialogs" on Windows (still never seen one). It adds a lot of complexity.

👍 x 💯;  Yes, it tangles up setting up various SDL states (and display modes) even before it's done parsing all the user inputs and the config file; and in doing so enforces an order-of-operations on these entangled activities. 

It will be so much cleaner to digest the conf and args in one pass, come up with our desired state (without any SDL code needed); and then setup SDL-once-and-only-once, and then launch the commands.

One side-thought is the SDL Mapper UI could be broken out into a completely separate binary that simply generates the mapper file; this will eliminate all it's runtime complexity and SDL-mode stuff in the code.

It's not like the mapper GUI is an ongoing part of routine game-play. We could still have a hot-key launch the executable (allowing the mapper to sit along-side the game), and we could have a hot-key that rereads mapper-file; or we auto-reload on timestamp changes.